CIP-XXX/CIP-XXX.md

```diff
@@ -0,0 +1,1408 @@
+```
+
+On the face of it, these schemes appear equivalent: they are all consistent, and
+all have formal descriptions, and quite similar ones at that. However, we
+believe that only the one we chose is the correct one. To explain this, we
+introduce two notions that we consider to be both intuitive and important,
+then specify why our choice of indexing scheme fits those notions better than
+any other.
+
+The first notion is _index locality_. Intuitively, this states that if
+two indexes are 'close' (in that their absolute difference is small), the values
+at those indexes should be 'close' (in that their positioning in memory should
+be separated less). We believe this notion to be reasonable, as this is an
+expectation from array indexing (and indeed, `BuiltinByteString` indexing), as
+well as the reason why packed array data is efficient on modern memory
+hierarchies. Extending this notion to bits, we can observe that the both-flip
+and no-flip variants of the bit indexing scheme do not preserve index locality:
+the separation between a bit at index $0$ and index $1$ is _significantly_
+different to the separation between a bit at index $7$ and index $8$ in both
+representations, despite their absolute difference being identical. Thus, we
+believe that these two variants are not viable, as they are not only confusing
+from the point of view of behaviour, they would also make implementation of
+future operations (such as shifts or rotations) significantly harder to both do,
+and also reason about. Thus, only the bit-flip variant, as well as our choice,
+remain contenders.
+
+The second notion is _conversion agreement_. To describe this notion, we first
+observe that, according to the definition of the bit indexing scheme given [in
+the corresponding section](#bit-indexing-scheme), as well as the corresponding
+definition for the bit-flip variant, we view a `BuiltinByteString` of length $n$
+as a binary natural number with exactly $8n$ digits, and the value at index $i$
+corresponds to the digits whose place value is either $2^i$ (for the bit-flip
+variant), or $2^{8n - i - 1}$ (for our chosen method). Put another way, under
+the specification for the bit-flip variant, the least significant binary digit
+is first, whereas in our chosen specification, the least significant binary
+digit is last. 
+
+When viewed this way, we can immediately see a potential problem, as by indexing
+a `BuiltinByteString`, we get back a `BuiltinInteger`, which has a numerical
+value as a natural number in the range $[0, 255]$. It would thus be sensible
+that, given a `BuiltinByteString` that is non-empty, if we were to get the
+values at bit indexes $0$ through $7$, and treat them as their corresponding
+place value in a summation, we should obtain the same answer as indexing
+whichever byte those bits come from. We call this notion _conversion agreement_,
+due to its relation to [CIP-121][conversion-cip], which allows us to
```

Comment:
CIP-121 allows for different endianness choices, does that not also matter here?

Reply:
It matters for one of the choices, namely the big-endian one. I should state this more clearly.

Reply:
It matters for one of the choices, namely the big-endian one. I should state this more clearly.